File: templates/mutation-incremental/test/diff.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { lineDiff, hunkIntersects, splitLines, normalizeText, MAX_EDIT_DISTANCE } from '../lib/diff.mjs';

const vectors = JSON.parse(readFileSync(new URL('../../../testdata/mutation-incremental/diff-vectors.json', import.meta.url), 'utf8'));

test('diff vectors', () => {
  assert.equal(vectors.maxEditDistance, MAX_EDIT_DISTANCE);
  for (const v of vectors.cases) assert.deepEqual(lineDiff(v.old, v.new).hunks, v.hunks, v.name);
});

test('normalisation and line splitting', () => {
  assert.equal(normalizeText('\ufeffa\r\nb\rc'), 'a\nb\nc');
  assert.deepEqual(splitLines(''), []);
  assert.deepEqual(splitLines('a\n\nb\n'), ['a', '', 'b']);
});

test('oldToNew maps unchanged lines across an insertion', () => {
  const { oldToNew } = lineDiff('a\nb\nc\n', 'x\na\nb\nc\n');
  assert.deepEqual(oldToNew, [null, 2, 3, 4]);
  const del = lineDiff('a\nb\nc\n', 'a\nc\n');
  assert.deepEqual(del.oldToNew, [null, 1, null, 2]);
});

test('hunkIntersects: overlap for a changed range, containment for an insertion point', () => {
  const changed = { oldStart: 3, oldEnd: 3, newStart: 3, newEnd: 3 };
  assert.equal(hunkIntersects(1, 5, changed), true);
  assert.equal(hunkIntersects(2, 2, changed), false);
  const insertAfter3 = { oldStart: 4, oldEnd: 3, newStart: 4, newEnd: 4 };
  assert.equal(hunkIntersects(1, 5, insertAfter3), true);
  assert.equal(hunkIntersects(1, 3, insertAfter3), false);
  assert.equal(hunkIntersects(4, 5, insertAfter3), false);
});

test('above MAX_EDIT_DISTANCE lineDiff returns null (the planner treats the file as whole)', () => {
  const a = Array.from({ length: MAX_EDIT_DISTANCE + 1 }, (_, i) => `a${i}`).join('\n');
  const b = Array.from({ length: MAX_EDIT_DISTANCE + 1 }, (_, i) => `b${i}`).join('\n');
  assert.equal(lineDiff(a, b), null);
});

File: internal/mutationfresh/diff_test.go
package mutationfresh

import (
	"encoding/json"
	"fmt"
	"os"
	"reflect"
	"strings"
	"testing"
)

// The harness runs the same vectors (templates/mutation-incremental/lib/diff.mjs), so the gate and
// the planner agree on which lines an edit touched.
func TestLineDiffVectors(t *testing.T) {
	data, err := os.ReadFile("../../testdata/mutation-incremental/diff-vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var vectors struct {
		MaxEditDistance int `json:"maxEditDistance"`
		Cases           []struct {
			Name  string `json:"name"`
			Old   string `json:"old"`
			New   string `json:"new"`
			Hunks []hunk `json:"hunks"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatal(err)
	}
	if vectors.MaxEditDistance != maxEditDistance {
		t.Fatalf("maxEditDistance %d, vectors say %d", maxEditDistance, vectors.MaxEditDistance)
	}
	for _, c := range vectors.Cases {
		got, ok := lineDiff(c.Old, c.New)
		if !ok || len(got)+len(c.Hunks) > 0 && !reflect.DeepEqual(got, c.Hunks) {
			t.Errorf("%s: got %+v (%v), want %+v", c.Name, got, ok, c.Hunks)
		}
	}
}

func TestLineDiffGivesUpAboveTheEditDistance(t *testing.T) {
	var a, b strings.Builder
	for i := 0; i <= maxEditDistance; i++ {
		fmt.Fprintf(&a, "a%d\n", i)
		fmt.Fprintf(&b, "b%d\n", i)
	}
	if _, ok := lineDiff(a.String(), b.String()); ok {
		t.Error("a diff above the edit distance must give up")
	}
}

// Spec §11.4.3: ranges overlap for a changed hunk; a pure insertion intersects a mutant that spans
// the insertion point.
func TestHunkIntersects(t *testing.T) {
	for _, c := range []struct {
		start, end int
		h          hunk
		want       bool
	}{
		{1, 5, hunk{OldStart: 2, OldEnd: 2, NewStart: 2, NewEnd: 2}, true},
		{3, 3, hunk{OldStart: 2, OldEnd: 2, NewStart: 2, NewEnd: 2}, false},
		{1, 5, hunk{OldStart: 6, OldEnd: 5, NewStart: 6, NewEnd: 6}, false},
		{1, 5, hunk{OldStart: 3, OldEnd: 2, NewStart: 3, NewEnd: 3}, true},
	} {
		if got := hunkIntersects(c.start, c.end, c.h); got != c.want {
			t.Errorf("[%d,%d] %+v = %v", c.start, c.end, c.h, got)
		}
	}
}
